src/DotNet.Status.Web/Controllers/AnnotationsController.cs

```diff
@@ -0,0 +1,109 @@
+
+namespace DotNet.Status.Web.Controllers
+{
+    /// <summary>
+    /// This Annotations controller serves queries in a format compatible with the Grafana
+    /// "Simple JSON Datasource". It is used to expose the "Deployments" table data to 
+    /// Grafana, which may then render the information in dashboards.
+    /// </summary>
+    [Route("api/[controller]")]
+    [ApiController]
+    public class AnnotationsController : ControllerBase
+    {
+        private const int _maximumServerCount = 10; // Safety limit on query complexity
+        private readonly ILogger<AnnotationsController> _logger;
+        private readonly IOptionsMonitor<AnnotationsOptions> _options;
+
+        public AnnotationsController(ILogger<AnnotationsController> logger, IOptionsMonitor<AnnotationsOptions> options)
+        {
+            _logger = logger;
+            _options = options;
+        }
+
+        [HttpGet]
+        public ActionResult Get()
+        {
+            return Ok();
+        }
+
+        [HttpPost]
+        [Route("query")]
+        public ActionResult Query()
+        {   // Required endpoint, but not used
+            return NoContent();
+        }
+
+        [HttpPost]
+        [Route("search")]
+        public ActionResult Search()
+        {   // Required endpoint, but not used
+            return NoContent();
+        }
+
+        [HttpPost]
+        [Route("annotations")]
+        public async IAsyncEnumerable<AnnotationEntry> Post(AnnotationQueryBody annotationQuery, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            TableClient tableClient = new TableClient(new Uri(_options.CurrentValue.DeploymentsTableConnectionString));
+
+            IEnumerable<string> services = (annotationQuery.Annotation.Query?.Split(',') ?? Array.Empty<string>())
+               .Take(_maximumServerCount)
+               .Where(s => !string.IsNullOrWhiteSpace(s))
+               .Select(s => s.Trim());
+
+            StringBuilder filterBuilder = new StringBuilder();
+            filterBuilder.Append($"Started gt datetime'{annotationQuery.Range.From:O}' and Ended lt datetime'{annotationQuery.Range.To:O}'");
+            if (services.Any())
+            {
+                filterBuilder.Append(" and (");
+                filterBuilder.Append(string.Join(" or ", services.Select(s => $"PartitionKey eq '{s}'")));
+                filterBuilder.Append(')');
+            }
+
+            string filter = filterBuilder.ToString();
+            _logger.LogTrace("Compiled filter query: {Query}", filter);
+
+            IAsyncEnumerable<DeploymentEntity> entityQuery = tableClient.QueryAsync<DeploymentEntity>(
+                filter: filter,
+                cancellationToken: cancellationToken);
+
+            await foreach (DeploymentEntity entity in entityQuery)
+            {
+                if (entity.Started == null)
+                {
+                    // Invalid entity, skip it.
+                    continue;
```

Comment:
We "lose" the calls a lot... do we want to maybe treat this like the started but not ended one below? At least some hint that _some_ sort of deployment happened. :-)

Reply:
Hm, yeah. Let me think a bit about that. I suppose if nothing else it can just be a "started" like you mentioned.